src/util/util.h

```diff
@@ -478,6 +478,31 @@ errno_t sss_filter_sanitize_for_dom(TALLOC_CTX *mem_ctx,
                                     char **sanitized,
                                     char **lc_sanitized);
 
+/* Sanitize an input string (e.g. a DN) for use in
+ * an LDAP/LDB filter
+ *
+ * It is basically the same as sss_filter_sanitize(_ex),
+ * just extra spaces inside DN around '=' and ',' are removed
+ * before sanitizing other characters . According the documentation
```

Comment:
Hi,

which documentation are you referring to? I found RFC 4514 which is quite vague here. What I got from it was that a space might be allowed after a ',' (RDN separator) to make the DN more readable. This would only allow trailing spaces in the attribute name.

For attribute values a leading or trailing space can be part of the attribute value if it is escaped. I would suggest to add this case to the tests as well.

bye,
Sumit

Reply:
Hi @sumit-bose, 

I found nice summary here https://ldap.com/ldap-dns-and-rdns/ (at the bottom)
This is defined in https://tools.ietf.org/html/rfc1779 (A String Representation of Distinguished Names).

This RFC clearly allows spaces around separators (see chapter https://tools.ietf.org/html/rfc1779#section-3).

I'm not 100% sure whether it is allowed in LDAP server reply, but it seems probable. I managed to reproduce this case with OpenLDAP so seems more generic problem.

Reply:
Hi @thalman,

thanks for the links. Please note that https://tools.ietf.org/html/rfc1779 is for LDAP v2, the current version is v3 and v2 was retired in 2003 by https://tools.ietf.org/html/rfc3494.

There is also an interesting section in RFC-2253 https://tools.ietf.org/html/rfc2253#section-4 where is is described what a client must allow when talking to LDAP v2 but also it is said that `Implementations MUST NOT, however, generate any of the RFC 1779 encodings which are not described above in section 2.`

About LDAP server reply. The LDAP server sends in its reply what the user has added e.g. in the `member` attribute. If the value was added with spaces and the OpenLDAP server does not sanitize the input is it imo expected that the reply will contain the spaces as well. I tried this with 389ds and here the input is sanitized automatically, no matter how many spaces I added to the attribute value in the ldapmodify request 389ds always store the value with all the unexpected spaces removed.

Just to make it clear, I'm not against adding this patch, but I want to make clear that we try to mitigate an issue which should be imo better be fixed on the server side by removing the spaces.

bye,
Sumit

Reply:
Just small clarification about openldap - openldap also sanitizes the input and DN is stripped of extra spaces. I achieved this by directly modifying the underlying database. Output is not sanitized and that allowed me to reproduce the issue.

Reply:
Hi,

I got confirmation from LDAP developer Marc Sauton and the spaces are valid and covered by RFC4514 Appendix A. So let's go ahead with this PR.

bye,
Sumit